### PR TITLE
fix IO deadlock condition

### LINKED ISSUE #22832: Deadlock in reading stdout from cmd
I want to read the stdout from an external program and process the results in Julia. However it sometimes hangs indefinitely.

It was previously [discussed on Discourse](https://discourse.julialang.org/t/deadlock-in-reading-stdout-from-cmd/4788/), but @vtjnash suggested to make an issue here, he could reproduce it by running this on all cores (I also see it on one core):

```julia
while true
    readstring(`dd if=/dev/zero bs=1048576 count=2`)
end
```

It just reads in two MiB of zeroes, in two blocks.
Successful iterations of the loop will print this to stderr:
```text
2+0 records in
2+0 records out
2097152 bytes (2,1 MB, 2,0 MiB) copied, 0,00447523 s, 469 MB/s
```

Until it hangs. Killing it with Ctrl-C then gives the following:
```text
1+0 records in
0+0 records out
0 bytes copied, 11,1429 s, 0,0 kB/s

signal (2): Interrupt
while loading /home/martijn/test/dd.jl, in expression starting on line 1
unknown function (ip: 0x7fb138855498)
uv__epoll_wait at /home/martijn/src/julia/deps/srccache/libuv-52d72a52cc7ccd570929990f010ed16e2ec604c8/src/unix/linux-syscalls.c:321
uv__io_poll at /home/martijn/src/julia/deps/srccache/libuv-52d72a52cc7ccd570929990f010ed16e2ec604c8/src/unix/linux-core.c:267
uv_run at /home/martijn/src/julia/deps/srccache/libuv-52d72a52cc7ccd570929990f010ed16e2ec604c8/src/unix/core.c:354
process_events at ./libuv.jl:82 [inlined]
wait at ./event.jl:216
wait at ./event.jl:27
jl_call_fptr_internal at /home/martijn/src/julia/src/julia_internal.h:369 [inlined]
jl_call_method_internal at /home/martijn/src/julia/src/julia_internal.h:388 [inlined]
jl_apply_generic at /home/martijn/src/julia/src/gf.c:1929
jl_apply_2va at /home/martijn/src/julia/src/rtutils.c:287
stream_wait at ./stream.jl:42
close at ./stream.jl:326
open at ./process.jl:588
read at ./process.jl:645
readstring at ./process.jl:651
macro expansion at /home/martijn/test/dd.jl:2 [inlined]
anonymous at ./<missing> (unknown line)
jl_call_fptr_internal at /home/martijn/src/julia/src/julia_internal.h:369 [inlined]
jl_call_method_internal at /home/martijn/src/julia/src/julia_internal.h:388 [inlined]
jl_toplevel_eval_flex at /home/martijn/src/julia/src/toplevel.c:594
jl_parse_eval_all at /home/martijn/src/julia/src/ast.c:909
jl_load at /home/martijn/src/julia/src/toplevel.c:620
include_from_node1 at ./loading.jl:556
jl_call_fptr_internal at /home/martijn/src/julia/src/julia_internal.h:369 [inlined]
jl_call_method_internal at /home/martijn/src/julia/src/julia_internal.h:388 [inlined]
jl_apply_generic at /home/martijn/src/julia/src/gf.c:1929
include at ./sysimg.jl:14
jl_call_fptr_internal at /home/martijn/src/julia/src/julia_internal.h:369 [inlined]
jl_call_method_internal at /home/martijn/src/julia/src/julia_internal.h:388 [inlined]
jl_apply_generic at /home/martijn/src/julia/src/gf.c:1929
process_options at ./client.jl:315
_start at ./client.jl:383
jl_call_fptr_internal at /home/martijn/src/julia/src/julia_internal.h:369 [inlined]
jl_call_method_internal at /home/martijn/src/julia/src/julia_internal.h:388 [inlined]
jl_apply_generic at /home/martijn/src/julia/src/gf.c:1929
unknown function (ip: 0x401e40)
unknown function (ip: 0x401602)
__libc_start_main at /build/glibc-bfm8X4/glibc-2.23/csu/../csu/libc-start.c:291
unknown function (ip: 0x4016b8)
unknown function (ip: 0xffffffffffffffff)
Allocations: 210788 (Pool: 210123; Big: 665); GC: 35
```

As noted on Discourse, when using smaller block sizes, it always seems to hang after 65536 (2^16) bytes. This is seen on  julia 0.5/0.6/master and all OS, though on Unix it seems to happen more quickly than Windows.

@vtjnash said that it seems like something is dropping an event notification.

cc @evetion

### PR DESCRIPTION
**Pull request by vtjnash**
when calling uv_shutdown on a handle being written to by another process
we might never get the UV__POLLOUT notification
but we also don't need to delay the uv_close call
if we have nothing to write

in the future, we should introduce a new `uv_drain` callback,
instead of continuing to abuse uv_shutdown for this purpose

fix #22832

### CODE REVIEW

File: src/jl_uv.c [hunk 1]
```diff
@@ -197,16 +197,18 @@ JL_DLLEXPORT void jl_close_uv(uv_handle_t *handle)
...
-        if (((uv_stream_t*)handle)->shutdown_req) {
+        if (stream->shutdown_req) {
 #endif
             // don't close the stream while attempting a graceful shutdown
             return;
         }
-        if (uv_is_writable((uv_stream_t*)handle)) {
+        if (uv_is_writable(stream) && stream->write_queue_size != 0) {
             // attempt graceful shutdown of writable streams to give them a chance to flush first
+            // TODO: introduce a uv_drain cb API instead instead of abusing uv_shutdown in this way
```
Comment (tkelman) [inline src/jl_uv.c]: duplicated "instead"

File: test/spawn.jl [hunk 2]
```diff
@@ -469,3 +475,18 @@ let c = `ls -l "foo bar"`
...
+    Base.link_pipe(stdin, julia_only_write=true)
+    p = spawn(pipeline(catcmd, stdin=stdin, stdout=stdout, stderr=DevNull))
+    @async begin # feed cat with 2 MB of data (zeros)
+        write(stdin, zeros(UInt8, 1048576 * 2))
+        close(stdin)
+    end
+    sleep(0.5) # give cat a chance to fill the write buffer for stdout
+    close(stdout.in) # make sure we can still close the write end
+    @test sizeof(readstring(stdout)) == 1048576 * 2 # make sure we get all the data
+    success(p)
```
Comment (tkelman) [inline test/spawn.jl]: `@test` ?

File: test/spawn.jl [hunk 3]
```diff
@@ -469,3 +479,19 @@ let c = `ls -l "foo bar"`
     @test length(c) == 3
     @test eachindex(c) == 1:3
 end
+
+## Deadlock in spawning a cmd (#22832)
+# FIXME?
```
Comment (tkelman) [inline test/spawn.jl]: what was broken about this? use test_broken

Comment (tkelman) [inline test/spawn.jl]: hello?

Comment (vtjnash) [inline test/spawn.jl]: Mentioned in the merge commit text. I don't know the answers to any of those questions, so I can't provide more information. If I knew any of those answers, I would have already fixed the test.

Comment (tkelman) [inline test/spawn.jl]: Would test_broken have worked? A simple response to the question "it froze on CI and couldn't reproduce locally" would be a lot more visible than the merge commit text.

**Pull request merged**